core/database/src/main/java/com/pronoidsoftware/core/database/RoomLocalAgendaDataSource.kt

```diff
@@ -59,4 +67,69 @@ class RoomLocalAgendaDataSource @Inject constructor(
     override suspend fun deleteAllReminders() {
         agendaDao.deleteAllReminders()
     }
+
+    // Tasks
+    override fun getAllTasks(): Flow<List<AgendaItem.Task>> {
+        return agendaDao.getAllTasks()
+            .map { taskEntities ->
+                taskEntities.map { it.toTask() }
+            }
+    }
+
+    override fun getTasksForDate(targetDate: String): Flow<List<AgendaItem.Task>> {
+        return agendaDao.getTasksForDate(targetDate)
+            .map { taskEntities ->
+                taskEntities.map { it.toTask() }
+            }
+    }
+
+    override suspend fun upsertTask(task: AgendaItem.Task): Result<TaskId, DataError.Local> {
+        return try {
+            val entity = task.toTaskEntity()
+            agendaDao.upsertTask(entity)
+            Result.Success(entity.id)
+        } catch (e: SQLiteFullException) {
+            Result.Error(DataError.Local.DISK_FULL)
+        }
+    }
+
+    override suspend fun upsertTasks(
+        tasks: List<AgendaItem.Task>,
+    ): Result<List<TaskId>, DataError.Local> {
+        return try {
+            val entities = tasks.map { it.toTaskEntity() }
+            agendaDao.upsertTasks(entities)
+            Result.Success(entities.map { it.id })
+        } catch (e: SQLiteFullException) {
+            Result.Error(DataError.Local.DISK_FULL)
+        }
+    }
+
+    override suspend fun deleteTask(id: String) {
+        agendaDao.deleteTask(id)
+    }
+
+    override suspend fun deleteAllTasks() {
+        agendaDao.deleteAllTasks()
+    }
+
+    // All
+    override fun getAllAgendaItems(): Flow<List<AgendaItem>> {
+        return getAllReminders()
+            .combine(getAllTasks()) { reminders, tasks ->
+                (reminders + tasks).sortedBy { it.startDateTime }
+            }
+    }
+
+    override fun getAgendaItemsForDate(targetDate: String): Flow<List<AgendaItem>> {
+        return getRemindersForDate(targetDate)
+            .combine(getTasksForDate(targetDate)) { reminders, tasks ->
+                (reminders + tasks).sortedBy { it.startDateTime }
+            }
+    }
+
+    override suspend fun deleteAllAgendaItems() {
+        agendaDao.deleteAllReminders()
+        agendaDao.deleteAllTasks()
```

Comment:
Here, I would recommend to use a database transaction instead. That will ensure that all these DB calls will only be applied to the DB if they all succeed. If one of these wouldn't succeed for whatever reason (out of storage, coroutine gets cancelled, etc.), your DB will be in an inconsistent state. The first calls may have already succeeded, but the later calls will never be executed due to a cancellation for example.
In Room, you can create a transaction (just as an example, needs to be adjusted to your use case):
```
@Dao
interface AgendaDao {

    @Transaction
    suspend fun deleteAllAgendaItems(db: AgendaDatabase) {
        db.apply {
            eventDao.deleteAllEvents()
            attendeeDao.deleteAllAttendees()
            taskDao.deleteAllTasks()
        }
    }
}
```
Whenever you need to execute multiple DB calls that should only be applied to the DB if they all succeed, you need something like this.

Reply:
Thanks for pointing this out, makes sense